pkg/token/token.go

```diff
@@ -210,23 +266,56 @@ func (g generator) GetWithRoleForSession(clusterID string, roleARN string, sess
 		stsAPI = sts.New(sess, &aws.Config{Credentials: creds})
 	}
 
-	return g.GetWithSTS(clusterID, stsAPI)
+	return g.GetWithSTS(clusterID, roleARN, stsAPI)
 }
 
 // GetWithSTS assumes returns a token valid for clusterID using the given STS client.
-func (g generator) GetWithSTS(clusterID string, stsAPI *sts.STS) (string, error) {
+func (g generator) GetWithSTS(clusterID string, roleARN string, stsAPI *sts.STS) (string, error) {
 	// generate an sts:GetCallerIdentity request and add our custom cluster ID header
 	request, _ := stsAPI.GetCallerIdentityRequest(&sts.GetCallerIdentityInput{})
 	request.HTTPRequest.Header.Add(clusterIDHeader, clusterID)
 
+	expiryTime := time.Minute
+
+	// if we're using a cache, keep the token valid for an hour (like STS does by default)
+	if g.useCache {
+		expiryTime = time.Hour
+	}
+
+	expiryDate := time.Now().Add(expiryTime)
+
 	// sign the request
-	presignedURLString, err := request.Presign(60 * time.Second)
+	presignedURLString, err := request.Presign(expiryTime)
 	if err != nil {
 		return "", err
 	}
 
 	// TODO: this may need to be a constant-time base64 encoding
-	return v1Prefix + base64.RawURLEncoding.EncodeToString([]byte(presignedURLString)), nil
+	tok := v1Prefix + base64.RawURLEncoding.EncodeToString([]byte(presignedURLString))
+
+	if g.useCache {
```

Comment:
Since I don't think it makes sense to cache tokens generated from users, only assumed roles, there should be a check that useCache is only set when a role is passed.

Reply:
Personally I believe that caching tokens generated from users does make sense in the MFA case, because even if you don't assume a role you're going to be prompted for credentials, and we'd like to avoid having to do that for every kubectl call.

Reply:
Makes sense.